tls/s2n_x509_validator.c

```diff
@@ -549,3 +555,71 @@ s2n_cert_validation_code s2n_x509_validator_validate_cert_stapled_ocsp_response(
     return ret_val;
 #endif /* S2N_OCSP_STAPLING_SUPPORTED */
 }
+
+DEFINE_POINTER_CLEANUP_FUNC(STACK_OF(X509)*, wipe_cert_chain);
+
+S2N_RESULT s2n_x509_validator_validate_certificate_signatures(struct s2n_connection *conn, struct s2n_x509_validator *validator,
+                                                                             s2n_cert_validation_code *validation_code)
+{
+    ENSURE_REF(conn);
+    ENSURE_REF(validator);
+
+    if (conn->config->security_policy->certificate_signature_preferences == NULL) {
+        *validation_code = S2N_CERT_OK;
+        return S2N_RESULT_OK;
+    }
+    
+    DEFER_CLEANUP(STACK_OF(X509) *validated_chain = X509_STORE_CTX_get1_chain(validator->store_ctx), wipe_cert_chain_pointer);
+
+    const int certs_in_chain = sk_X509_num(validated_chain);
+
+    /* Do not validate the root certificate */
+    unsigned int certs_to_validate = certs_in_chain - 1;
+    
+    for (size_t i = 0; i < certs_to_validate; i++) {
+        bool out = false;
+        X509 *cert = sk_X509_value(validated_chain, i);
+
+        GUARD_RESULT(s2n_is_certificate_sig_scheme_supported(conn, cert, conn->config->security_policy->certificate_signature_preferences, &out));
+        if(out == false) {
+            *validation_code = S2N_CERT_ERR_UNTRUSTED;
+            return S2N_RESULT_OK;
+        }
+    }
+
+    *validation_code = S2N_CERT_OK;
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_is_certificate_sig_scheme_supported(struct s2n_connection *conn, X509 *x509_cert, const struct s2n_signature_preferences *cert_sig_preferences, bool *out)
+{
+    ENSURE_REF(conn);
+    ENSURE_REF(x509_cert);
+    ENSURE_REF(cert_sig_preferences);
+    ENSURE_REF(out);
+
+    int nid = 0;
+
+    #if defined(LIBRESSL_VERSION_NUMBER) && (LIBRESSL_VERSION_NUMBER < 0x02070000f)
```

Comment:
How have you tested that this is correct?

Reply:
This passes our integ tests for version 2.6.4(our current libressl version) and I have manually tested version 2.7.2, which is what this number is referencing. Is there any other way of testing this?

Reply:
I can't think of any better way to test this than manual testing against different libcryptos. And the documentation says that the APIs should be available in all versions of Openssl and BoringSSL?

Reply:
Is the preference to use `X509_get_signature_nid` and only for LIBRESSL_VERSIONS  < 2.7.2 use the `OBJ_obj2nid` API?

Reply:
To summarize:
X509_get_signature_nid is supported in Openssl 1.0.2 and above, which is the first Openssl version s2n supports. It is supported in all versions of boringssl and it was added in Libressl in 2.7.0, which is 0x02070000f above. Since we support libressl 2.6.4, we have to use an ifdef for this to work.